File: 647/Palindromic_Substrings.c
/*
//dp
int countSubstrings(char* s) {
    if (!s) return 0;
    int len = strlen(s);
    if (len == 0) return 0;
    int **d = (int **)malloc(len * sizeof(int *));
    if (!d) return 0;

    int i, j, count = 0;
    for (i = 0; i < len; i++) {
        d[i] = (int *)calloc(len, sizeof(int));
        if (!d[i]) return 0;
    }

    for (i = len - 1; i >= 0; i--) {
        for (j = i; j < len; j++) {
            if (j == i) d[i][j] = 1;
            else if (j == i+1 && s[j] == s[i]) d[i][j] = 1;
            else {
                if (s[j] == s[i]) d[i][j] = d[i+1][j-1];
            }
            if (d[i][j] == 1) count++;
        }
    }

    for (i = 0; i < len; i++) {
        free(d[i]);
    }
    free(d);
    return count;
}
*/

//expand
int expand(char *s, int left, int right) {
    int count = 0, len = strlen(s);
    while (left >= 0 && right < len && s[left] == s[right]) {
        count++;
        left--;
        right++;
    }
    return count;
}

int countSubstrings(char* s) {
    if (!s) return 0;
    int count = 0, len = strlen(s), i;

    for (i = 0; i < len; i++) {
        count += expand(s, i, i);
        count += expand(s, i, i+1);
    }
    return count;
}

File: 647/Palindromic_Substrings.py
class Solution:
    def dp(self, s: str) -> int:
        length = len(s)
        res = 0

        d = [[False for x in range(length)] for y in range(length)]

        for x in range(length - 1, -1, -1):
            for y in range(x, length):
                if x == y:
                    d[x][y] = True
                elif x + 1 == y and s[y] == s[x]:
                    d[x][y] = True
                else:
                    d[x][y] = s[x] == s[y] and d[x + 1][y - 1]

                if d[x][y]:
                    res += 1

        return res

    def expand(self, s: str, left: int, right: int) -> int:
        length = len(s)
        res = 0

        while left >= 0 and right < length and s[left] == s[right]:
            res += 1
            left -= 1
            right += 1

        return res

    def countSubstrings(self, s: str) -> int:
        if not s or len(s) == 0:
            return 0

        # solution 1
        # return self.dp(s)

        # solution 2
        length = len(s)
        res = 0

        for x in range(length):
            res += self.expand(s, x, x)
            res += self.expand(s, x, x + 1)

        return res
